File: chat-backend/src/main/java/com/smallchat/backend/chat/domain/model/vo/Message.java
package com.smallchat.backend.chat.domain.model.vo;

import com.smallchat.backend.chat.framework.web.dto.MessageBasicInfo;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.ZonedDateTime;

@Getter
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "message")
public class Message {
    private String message;
    private String chatId;
    private Sender sender; //optional, can be nullable if messageType is SYSTEM...
    private ZonedDateTime createdAt;
    private MessageType messageType;

    public Message(String message, String chatId, Sender sender) {
        this.message = message;
        this.chatId = chatId;
        this.sender = sender;
        this.createdAt = ZonedDateTime.now();
        this.messageType = MessageType.USER;
    }

    public Message(SystemMessage message, String messageValue, String chatId) {
        this.message = message.getSystemMessage(messageValue);
        this.chatId = chatId;
        this.sender = null;
        this.createdAt = ZonedDateTime.now();
        this.messageType = MessageType.SYSTEM;
    }

    public static Message systemMessage(SystemMessage systemMessage, String messageValue, String chatId) {
        return new Message(systemMessage, messageValue, chatId);
    }

    public static Message notFoundMessage() {
        return new Message("정보 없음", "", null);
    }

    public MessageBasicInfo toMessageBasicInfo() {
        return new MessageBasicInfo(sender, createdAt, message, messageType);
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/chat/domain/model/Message.kt
package com.smallchat.backend.chat.domain.model

import org.springframework.data.mongodb.core.mapping.Document
import java.time.Instant

@Document(collection = "message")
data class Message(
    val chatId: String,
    val userId: String,
    val nickname: String,
    val message: String,

    val sentAt: Instant? = Instant.now(),
    val type: MessageType? = MessageType.USER
) {

    val actualSentAt: Instant get() = sentAt ?: Instant.now()

    companion object {
        fun userMessage(chatId: String, userId: String, nickname: String, message: String): Message {
            return Message(chatId, userId, nickname, message, Instant.now(), MessageType.USER)
        }

        fun systemJoin(chatId: String, userNickname: String): Message {
            return Message(
                chatId = chatId,
                userId = SystemMessageKt.SYSTEM_ID,
                nickname = SystemMessageKt.SYSTEM_NICKNAME,
                message = SystemMessageKt.userJoined(userNickname),
                type = MessageType.SYSTEM
            )
        }

        fun systemLeave(chatId: String, userNickname: String): Message =
            Message(
                chatId = chatId,
                userId = SystemMessageKt.SYSTEM_ID,
                nickname = SystemMessageKt.SYSTEM_NICKNAME,
                message = SystemMessageKt.userLeft(userNickname),
                type = MessageType.SYSTEM
            )

        fun systemCreated(chatId: String): Message =
            Message(
                chatId = chatId,
                userId = SystemMessageKt.SYSTEM_ID,
                nickname = SystemMessageKt.SYSTEM_NICKNAME,
                message = SystemMessageKt.chatCreated(),
                type = MessageType.SYSTEM
            )
    }
}